Kernel/FileSystem/FATFileSystem.cpp

```diff
@@ -0,0 +1,352 @@
+        if (entry->filename[0] == 0x00) {
+            dbgln_if(FAT_DEBUG, "FATFS: Found end entry");
+            return inodes;
+        } else if (static_cast<u8>(entry->filename[0]) == 0xE5) {
+            dbgln_if(FAT_DEBUG, "FATFS: Found unused entry");
+        } else if (entry->attributes == FATAttributes::LongFileName) {
+            dbgln_if(FAT_DEBUG, "FATFS: Found LFN entry");
+            lfn_entries.append(*reinterpret_cast<FATLongFileNameEntry*>(entry));
+        } else {
+            dbgln_if(FAT_DEBUG, "FATFS: Found 8.3 entry");
+            lfn_entries.reverse();
+            auto inode = TRY(FATInode::create(fs(), *entry, lfn_entries));
+            inodes.append(inode);
+            lfn_entries.clear();
+        }
+    }
+
+    return EINVAL;
+}
+
+ErrorOr<NonnullOwnPtr<KString>> FATInode::compute_filename(FATEntry& entry, Vector<FATLongFileNameEntry> const& lfn_entries)
+{
+    if (lfn_entries.is_empty()) {
+        StringBuilder filename;
+        filename.append(byte_terminated_string(StringView(entry.filename, 8), ' '));
+        if (entry.extension[0] != ' ') {
+            filename.append('.');
+            filename.append(byte_terminated_string(StringView(entry.extension, 3), ' '));
+        }
+        return TRY(KString::try_create(filename.string_view()));
+    } else {
+        StringBuilder filename;
+        for (auto& lfn_entry : lfn_entries) {
+            filename.append(lfn_entry.characters1[0]);
+            filename.append(lfn_entry.characters1[1]);
+            filename.append(lfn_entry.characters1[2]);
+            filename.append(lfn_entry.characters1[3]);
+            filename.append(lfn_entry.characters1[4]);
+            filename.append(lfn_entry.characters2[0]);
+            filename.append(lfn_entry.characters2[1]);
+            filename.append(lfn_entry.characters2[2]);
+            filename.append(lfn_entry.characters2[3]);
+            filename.append(lfn_entry.characters2[4]);
+            filename.append(lfn_entry.characters2[5]);
+            filename.append(lfn_entry.characters3[0]);
+            filename.append(lfn_entry.characters3[1]);
+        }
+
+        return TRY(KString::try_create(byte_terminated_string(filename.string_view(), 0xFF)));
+    }
+
+    VERIFY_NOT_REACHED();
+}
+
+time_t FATInode::fat_date_time(u16 date, u16 time) const
+{
+    time_t year = (date >> 9) + (1980 - 1970);
+    time_t month = (date >> 5) & 0b1111;
+    time_t day = date & 0b11111;
+
+    time_t hour = time >> 11;
+    time_t minute = (time >> 5) & 0b1111111;
+    time_t second = (time & 0b11111) << 1;
+
+    return second + minute * 60 + hour * 3600
+        + day + (24 * 3600) + month * (30 * 24 * 3600) + year * (365 * 24 * 3600);
+}
+
+// TODO: This might be worth putting as a part of the StringView class
+StringView FATInode::byte_terminated_string(StringView string, u8 fill_byte)
+{
+    while (string.length() > 0) {
```

Comment:
It's very hard to figure out what this algorithm is trying to do.

It *looks* like you're doing some kind of `find_last_not_of`?

```
Optional<size_t> find_last_not_of(StringView haystack, char needle)
{
    for (size_t i = haystack.length(); i > 0; --i) {
        if (haystack[i - 1] != needle)
            return i - 1;
    }
    return {};
}


StringView FATInode::byte_terminated_string(StringView string, u8 fill_byte)
{
    if (auto index = find_last_not_of(string, fill_byte); index.has_value())
         return string.substring_view(0, index.value();
    return string;
}
```